### PR TITLE
reverse_string.py

### PR DESCRIPTION
**Pull request by amymhaddad**

### CODE REVIEW

File: reverse_string.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+def reverse(text):
+    """Reverse a string"""
+
+    if text == "":
+        return text
+    else:
+        text_list = list(text)
+        text_list.reverse()
+        new_word = "".join(text_list)
```
Comment (paulghaddad) [inline reverse_string.py]: `reversed_string` is a more precise name.

File: reverse_string.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+"""Tests to reverse various strings"""
+
+def reverse(text):
+    """Reverse a string"""
+
+    if text == "":
+        return text
+    else:
+        text_list = list(text)
```
Comment (paulghaddad) [inline reverse_string.py]: Excellent strategy here! There are many ways to do this, and you chose the best way in my opinion.

**Comment by amymhaddad** [comment]
@paulghaddad I updated this exercise.

File: reverse_string.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+"""Tests to reverse various strings"""
+
+def reverse(text):
+    """Reverse a string"""
+
+    if text == "":
+        return text
+    else:
```
Comment (paulghaddad) [inline reverse_string.py]: So here's another example where you do a return in the `if`. This is called a "guard" clause, which basically checks for a strange edge case, in this example an empty string and returns the necessary value.

So one thing you can do here is sort of separate the `if` clause from the `else` clause. And as we discussed this morning, you can essentially get rid of the `else` clause and have the code in it simply run after the the guard clause check in the `if`. So something like this:

```
if text == "":
  return text

text_list = list(text)
text_list.reverse()
reversed_string = "".join(text_list)

return reversed_string
```

Note how the majority of the cases should not hit the guard clause and this visually separates the guard clause from the code that handles the majority of the work. It is _guarding_ the code that actually reverses the string.

**Comment by amymhaddad** [comment]
@paulghaddad Great info to have! I updated the exercise.

**Comment by paulghaddad** [comment]
@amymhaddad Excellent. This looks good. You're solid on approach #1. Now, comment out that approach and try a second, this time using a loop instead of the built-in `reverse` method. This will give you an opportunity to use loops and think of an algorithm.

**Comment by amymhaddad** [comment]
@paulghaddad I added a second option to this file using a loop.

File: reverse_string.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+#Option 2:
+"""Tests to reverse various strings"""
+
+def reverse_string(word):
+    """Reverse a given string"""
+
+    new_word = ''
+
+    for i in reversed(range(len(word))):
```
Comment (paulghaddad) [inline reverse_string.py]: This is awesome! I didn't know about `reversed`. Nice job coming up with the algorithm: Iterating in reverse through the original string and building up `new_word`!

File: reverse_string.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        reversed_word = word[i]
+        new_word += reversed_word
+    return new_word
+
+WORD_IN_REVERSE = reverse_string('cool')
```
Comment (paulghaddad) [inline reverse_string.py]: Remove the test code on lines 31-33.

**Pull request merged**